File: q49.py
from typing import List
from collections import defaultdict

class Solution:
  def groupAnagrams(self, str: List[str]) -> List[List[str]]:
    res = defaultdict(list)

    for s in str:
      count = [0] * 26

      for c in s:
        count[ord(c) - ord("a")] += 1

      res[tuple(count)].append(s)

    return res.values()

s = Solution()
print(s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

File: q49.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const map = new Map();

  for (let i = 0; i < strs.length; i++) {
    const currentString = strs[i];
    const sortCurrentString = currentString.split("").sort().join("");

    if (map.has(sortCurrentString)) {
      const originalString = map.get(sortCurrentString);
      map.set(sortCurrentString, [currentString, ...originalString]);
    } else {
      map.set(sortCurrentString, [currentString]);
    }
  }

  return Array.from(map.values());
};

console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));
